### PR TITLE
sync service: migrate subtrees to redis streams.

### LINKED ISSUE #308: migrating sync subtrees to redis streams
We are finding some weirdness with Redis keyspace notifications. See https://github.com/ipfs/testground/issues/252. If we are willing to give up ZooKeeper-like ephemeral nodes, adopting Redis streams is a good option as it would greatly simplify the subtree watch/write logic of the sync service.

Related:

https://github.com/ipfs/testground/pull/263#issuecomment-565575281
https://github.com/ipfs/testground/pull/263
https://github.com/ipfs/testground/issues/27

### PR DESCRIPTION
**Pull request by raulk**
@hacdias and I pair programmed to land this migration.

This has been tested with up to 400 nodes with the `local:exec` runner, and 96 nodes with the `local:docker` runner. It's stable and fast, compared to the previous solution.

This PR supersedes #263.

Fixes #308.

### CODE REVIEW

**Comment by nonsense** [comment]
Haven't look in close detail, but overall looks good.

**Comment by hacdias** [comment]
Tried running the DHT/FIND-PEERS plan but something funky is happening: it seems to be aborting, but the testground command does not exit. And there's some deadline being exceeded which I do not know if it's caused by this PR or not. 

image

**Update:** just ran it again and it worked and it actually exited correctly. probably something with the DHT unrelated to this PR.

image

**Comment by raulk** [comment]
I’ve tested this with 500 local:exec instances, and I hit a race during subscription closure that makes things hang. I have a pretty good sense of where this is. Will fix tomorrow.

**Comment by raulk** [comment]
I've fixed the subscription closure logic, and now I'm able to run 500 and 600 instance tests with `local:exec` with no sync-service related errors. As a baseline, the DHT find-peers test case with 100 nodes completes in ~10s, 200 nodes completes in ~20s. That's pretty fast and enables really quick feedback loops.

Some instances in the DHT find-peers test case still error out with "routing: not found" frequently, but the sync service is working brilliantly now in my tests.

File: plans/dht/test/common.go [hunk 1]
```diff
@@ -92,6 +92,7 @@ func NewDHTNode(ctx context.Context, runenv *runtime.RunEnv, opts *SetupOpts) (h
 	dhtOptions := []dhtopts.Option{
 		dhtopts.Datastore(datastore.NewMapDatastore()),
 		dhtopts.BucketSize(opts.BucketSize),
+		dhtopts.RoutingTableRefreshQueryTimeout(opts.Timeout),
```
[the remark this answers is not in the document]
Comment (raulk) [reply plans/dht/test/common.go]: I don't think we can hard-code it; needs to be somehow relative to the test user-provided timeout with some kind of floor, maybe? Let's do in a separate PR.

**Comment by raulk** [comment]
If you're interested in seeing this in action with 600 instances (hope your computer can cope!)

```
⟩ ./testground -vv run dht/find-peers \
      --builder=exec:go \
      --runner=local:exec \
      --build-cfg bypass_cache=true \
      --instances=600 \
      --test-param timeout_secs=240 \
      --test-param auto_refresh=false
```

Note that you will see test errors -- they relate to the logic in the test, not the sync service.

**Pull request merged**